### DISCUSSION TITLE
Handle internally generated events like user input

### DISCUSSION THREAD

**Discussion by henrySol** [jspsych/jsPsych#3050] [Q&A] [answered]
Hi everyone, hope you are doing well!

I am currently trying to figure out a way to use events that are generated by my experiment script instead of user input. What I am looking for is something like this:

- An event is dispatched based on some condition, for example using a simple function like this:
`    function simulateKeyPress(keyCode) {
        var event = new KeyboardEvent('keydown', { keyCode: keyCode });
        document.dispatchEvent(event);
        };`

- A jspsych trial, for example a jsPsychHtmlKeyboardResponse trial, handles the event as if it was an actual key press by the user. The event should be handled like a user response.

I have already figured out how to set up event listeners that handle the dispatched events. But I cannot find a way to let the jspsych trial object handle the event as a user response/user input (instead of listening for actual key presses by the user). Perhaps I have to modify the relevant plugins? Any help would be greatly apreciated!

Thanks in advance and enjoy your days

Henry

**Comment by jodeleeuw** [comment, marked as answer]
Hi @henrySol 

You probably need to dispatch the event on the specific element that contains the jsPsych experiment. When plugins use `getKeyboardResponse` the event listener is attached to the same element that is accessed via `jsPsych.getDisplayContainerElement()`.

Something like the following should work:

```js
jsPsych.getDisplayContainerElement().dispatchEvent(event);
```

**Reply by henrySol** [reply]
Hey @jodeleeuw ,

this was exactly what I needed! Thanks a lot! This was the missing piece and now I will be able to handle signals that come in via a serial port as user input (key presses), which allows me to use jsPsych in an MRI scanner environment (which is great).
